### ISSUE TITLE
nearest-neighbor by default?

### ISSUE THREAD

**Issue by NEO-SPECTRUMAN** [Feodor2/Mypal68#89]
собственно сабж
можно ли включить именно такое масштабирование картинок по дефолту?
нагуглить ненагуглилось

**Comment by zanud** [comment]
А записать в userContent.css:
`* { image-rendering: crisp-edges !important; } `
не помогает?

**Comment by NEO-SPECTRUMAN** [comment]
щас попробуем

**Comment by NEO-SPECTRUMAN** [comment]
```
img {
  image-rendering: crisp-edges !important;
}
```

не помогает
![2022-07-09_211801](https://user-images.githubusercontent.com/88002487/178118152-d8b5e9eb-85bf-4b21-beb4-141d79f69068.png)

была бы гамма коррекция
оно бы так не бесило...

**Comment by NEO-SPECTRUMAN** [comment]
хотелось бы конечно оно с about:config-а...
но видимо такого нету

**Comment by NEO-SPECTRUMAN** [comment]
хотя есть подохрение что страницы могут все равно выставлять свой вариант

**Comment by zanud** [comment]
> во нашел в вебархиве https://web.archive.org/web/20201124105355/https://developer.mozilla.org/ru/docs/Web/CSS/image-rendering crisp-edges это мурня 

В текущей версии документации (https://developer.mozilla.org/ru/docs/Web/CSS/image-rendering) говорится, что `crisp-edges` - это nearest-neighbor в чистом виде, а `pixelated` - двухэтапная интерполяция: сначала вверх с помощью nearest-neighbor, а потом вниз, до нужного размера, с помощью билинейной интерполяции.
Почему между этими двумя версиями документации такая огромная разница в описании одних и тех же параметров, не знаю. Но предпочёл поверить текущей, а проверить было не на чем.

> хотелось бы конечно оно с about:config-а...
> но видимо такого нету

Но это же к CSS относится. Подобные параметры, по-моему, никогда через about:config не управлялись.

> хотя есть подозрение что страницы могут все равно выставлять свой вариант

Могут. Поэтому здесь в правило вписан довесок `!important`, что делает это правило более приоритетным, чем правило на странице.
Но если и у правила на странице будет `!important`, то соревнование выиграет оно.
Однако на страницах `!important` встречается крайне редко.

**Comment by NEO-SPECTRUMAN** [comment]
текущая веррсия говорит что page not found

**Comment by NEO-SPECTRUMAN** [comment]
> Почему между этими двумя версиями документации такая огромная разница в описании одних и тех же параметров, не знаю.

новая версия звучит как бред

**Comment by NEO-SPECTRUMAN** [comment]
щас еще проверю как работает сплющевание

**Comment by zanud** [comment]
> новая версия звучит как бред

Нет. Есть такой приём повышения качества интерполяции изображения: сначала разрешение очень сильно увеличивают ("раздувают"), а потом только уменьшают до нужного. Это работает, причём независимо от того, больше конечное разрешение исходного или меньше.
Но вот соответствие названий методов и их описаний - да, какое-то противоестественное. Впрочем, я с подобным сталкивался в официальном описании методов преобразования цветовых пространств Фотошопом (сначала на сайте Adobe писали одно, а через несколько лет совсем другое), так что сейчас не очень удивился.

**Comment by NEO-SPECTRUMAN** [comment]
я знаю про приемы
название не соответствует

**Comment by NEO-SPECTRUMAN** [comment]
а так проверил
на serpent-е оно не ссоответствует вдоль и попере ни тому ни тому описанию :)

нужно проверять все наши "старые" ХРю браузеры
и сверять с новомодными firefox100500

**Comment by zanud** [comment]
На нынешней страничке документации есть окошко, где можно посмотреть влияние этого параметра на увеличение изображения.
У Firefox 102 `crisp-edges` и `pixelated` дают одинаково чёткий результат.
У Mypal68 `crisp-edges` чёткий результат, а `pixelated` - сглаженный (что, как ни странно, соответствует документации :-))

А я перед этим как раз хотел написать, что разногласия у документации могут быть связаны с изменениями в самой программе.

**Comment by NEO-SPECTRUMAN** [comment]
> (что, как ни странно, соответствует документации :-))

это мало похоже на описанный алгоритм
и там всего 2 режима из 5
да и тест галимый

нужно писать свой тест из 3-х строчек.....

**Comment by NEO-SPECTRUMAN** [comment]
тот же быдло сайтек гговорит что pixelated и crisp-edges
были добавлены только в firefox 65...93 версиях :rofl:

**Comment by NEO-SPECTRUMAN** [comment]
давайте и @Feodor2 -у обратим внимание
что есть нестыковки в режимах  image-rendering: -а
просто "для общего развития"

**Comment by zanud** [comment]
> и там всего 2 режима из 5

А их реально столько и есть. "auto" не в счёт - это всего лишь автоматический выбор чего-то из имеющихся алгоритмов. А последние два из пяти - экспериментальные, в стандарт не входят.

> нужно писать свой тест из 3-х строчек.....

Ну, строчек получается побольше, чем 3...
Вот, сделал [тестик](http://89.184.66.94/image-rendering.html). С кратным увеличением/уменьшением (крайние столбцы) и некратным (средние столбцы).

В процессе обнаружилось, что у Mypal68 перечень режимов сильно отличается от того, что в документации написано. И, в частности, pixelated вообще отсутствует (я в таблицу его внёс, но выделил курсивом).

> давайте и @Feodor2 -у обратим внимание
> что есть нестыковки в режимах image-rendering: -а
> просто "для общего развития"

Разве что "для общего развития". Потому что Mypal68 для всего внешнего мира притворяется Firefox 68, и ломать совместимость с ним не следует.

К тому же "несоответствие" таковым кажется только нам. А на самом деле вот есть занимательное чтиво: [https://drafts.csswg.org/css-images-3/#the-image-rendering](https://drafts.csswg.org/css-images-3/#the-image-rendering)

**Comment by NEO-SPECTRUMAN** [comment]
> К тому же "нестыковки" таковыми кажутся только нам. А на самом деле вот есть занимательное чтиво: https://drafts.csswg.org/css-images-3/#the-image-rendering

черным по белому написано

> pixelated
>     The image is scaled in a way that preserves the pixelated nature of the original as much as possible, 

где тут хоть капля  pixelated nature?
![2022-07-10_104241](https://user-images.githubusercontent.com/88002487/178135883-f448df5a-401d-4d70-8e4c-7d003a681c02.png)

**Comment by NEO-SPECTRUMAN** [comment]
и для теста нужно растянуть что нибудь с х8 масштабированием

**Comment by NEO-SPECTRUMAN** [comment]
и еще из портянки 

> Note: At integer multiples of the natural size, this gives the same results as [crisp-edges](https://drafts.csswg.org/css-images-3/#valdef-image-rendering-crisp-edges)

**Comment by zanud** [comment]
> где тут хоть капля pixelated nature?

Я же написал, что в Mypal68 `pixelated` не поддерживается, и эта строчка в таблице чисто справочно - посмотреть, как другие браузеры её отрабатывают.
Более того, на мозилловской странице документации написано, что `pixelated` не поддерживается в Firefox любой версии:
> In Firefox, `crisp-edges` is interpreted as nearest-neighbor, `pixelated` is not supported, and `auto` is interpolated as trilinear or linear.

Правда, на той же страничке ровно абзацем выше в таблице совместимости значится, что `pixelated` поддерживается Firefox, начиная с версии 93.

> и для теста нужно растянуть что нибудь с х8 масштабированием как минимум

Да пожалуйста. Никто не мешает скачать эту страничку к себе и изменить в ней размеры как душе угодно.

> можно сразу делать как "правильно"

А "правильно" - это как? Так, как хочется, или так, как в стандарте написано?
В Мозилловском описании текст практически один-в-один перетянут из стандарта (ссылку я давал).

**Comment by NEO-SPECTRUMAN** [comment]
> А "правильно" - это как?

это противоположное к "не правильно" :rofl:

а еще местами упоминается
-moz-crisp-edges

и есть подобные команды для IE
через -ms
...

**Comment by zanud** [comment]
> а еще местами упоминается
> -moz-crisp-edges

Да, есть такое. Но браузероспецифичные варианты я предпочитаю не трогать.

**Comment by NEO-SPECTRUMAN** [comment]
это как раз браузероспецифические именно от нашего браузера

**Comment by NEO-SPECTRUMAN** [comment]
а где внятно написано что pixelated не пподдерживается?

**Comment by zanud** [comment]
> это как раз браузероспецифические именно от нашего браузера

А есть основания полагать, что в нём `-moz-crisp-edges` отличается от `crisp-edges`?
Вот если бы `-moz-pixelated` было - тогда другое дело.

**Comment by NEO-SPECTRUMAN** [comment]
> А есть основания полагать,

а есть основание слепо верить всему что написано на заборе?
я постоянно сталкиваюсь с тем что описание не соответствует реальности

**Comment by zanud** [comment]
> а есть основание слепо верить всему что написано на заборе?
> я постоянно сталкиваюсь с тем, что описание не соответствует реальности

Я тоже вынужден "жить и работать" в условиях, когда документации местами сильно не соответствует действительности или отсутствуют в принципе. Поэтому уже чутьё есть.

Но добавил в табличку ещё одну строку. И увеличение увеличил :-)

**Comment by NEO-SPECTRUMAN** [comment]
вот в serpent55
-moz-crisp-edges отработал как надо
а crisp-edges отработал как auto видимо за неимением

**Comment by zanud** [comment]
Набор поддерживаемых значений можно посмотреть в отладчике, попытавшись отредактировать правило в стиле элемента - там выпадает список из всех возможных вариантов.

**Comment by Feodor2** [comment]
Объясните на пальцах что не так и что вы хотите, кокретно сейчас сам вникать в это не хочу.

**Comment by zanud** [comment]
> Объясните на пальцах что не так и что вы хотите

Если браузеру нужно показать на страничке картинку не в масштабе 1:1, то он вынужден заниматься интерполяцией изображения. Алгоритмов интерполяции существует множество разных.

Создатели сайтов об этом, как правило, не подозревают, но CSS позволяет свойством `image-rendering` задать, какой алгоритм следует использовать. Названия там, правда, неестественные и выбор крайне невелик, да ещё и не все, предусмотренные стандартом, поддерживаются конкретным браузером. (Правда, браузер при этом может поддерживать и нестандартные/устаревшие).

Конкретно у Mypal68 список состоит из "auto", "crisp-edges", "-moz-crisp-edges", "optimizequality" и "optimizespeed". По умолчанию (если свойство `image-rendering` явно не прописано) используется метод "auto".

"Желание-минимум" @NEO-SPECTRUMAN: чтобы каким-то параметром в **about:config** можно было задать метод по умолчанию, отличный от "auto".
"Желание-максимум": чтобы метод, заданный этим параметром, применялся всегда, даже если в свойствах объекта DOM явно указан другой.

(Строго говоря, его интересует не любой из возможных методов интерполяции, а конкретно "nearest-neighbor interpolation", который в CSS носит название "crisp-edges".)

Способ обойтись без **about:config** я предложил, и он работает. Но это задание метода, используемого по умолчанию, оно не действует на изображения, у которых метод интерполяции задан явно через CSS.

**Comment by NEO-SPECTRUMAN** [comment]
> Алгоритмов интерполяции существует множество разных.

но основная проблема
что ни одна зараза обычно не преобразует srgb яркость в линейный вид для расчетов

**Comment by zanud** [comment]
> но основная проблема
> что ни одна зараза обычно не преобразует srgb яркость в линейный вид для расчетов

И переключение `gfx.color_management.mode` в значение **1** не помогает?

**Comment by NEO-SPECTRUMAN** [comment]
нет

**Comment by NEO-SPECTRUMAN** [comment]
это только для png-жей
с конкретном указанным профилем в формате

**Comment by zanud** [comment]
> это только для png-жей
> с конкретном указанным профилем в формате

1. Не только PNG, а любые поддерживаемые графические форматы.
2. **"2"** означает "картинки, содержащие цветовой профиль, выводить на экран, учитывая цветовые пространства".
**"1"**  означает "картинки выводить на экран, учитывая цветовые пространства, при этом картинки, не содержащие цветовой профиль, трактовать как находящиеся в SRGB".
3. Картинки с цветовым профилем в Интернете встречаются крайне редко. Нет, может на каких-то специализированных сайтах они содержатся в заметных количествах, но мне за всю жизнь таких файлов всего несколько штук попалось.
4. Официальное описание говорит об учёте профилей при выводе на экран, но ничего не говорит об их учёте при масштабировании и прочих преобразованиях изображения. Для них его трактовать можно по-разному.

**Comment by NEO-SPECTRUMAN** [comment]
> 4\. Официальное описание говорит об учёте профилей при выводе на экран, но ничего не говорит об их учёте при масштабировании и прочих преобразованиях изображения. Для них его трактовать можно по-разному.

IRL большинство изображений требует масштабироания для отображения
а вот профили могут ссохранять всякие камеры и сканеры
которые из покон веков делают картинки в несколько раз больше чем мониторы

**Comment by zanud** [comment]
> профили могут ссохранять всякие камеры и сканеры
> которые из покон веков делают картинки в несколько раз больше чем мониторы

Сканер просто "по определению" отдаёт изображение только без профиля.
Профессиональные фотоаппараты сохраняют снимки без профиля.
Камеры смартфонов (смотрел в файлах, валяющихся у меня на диске) сохраняют снимки без профиля.
Зато почему-то профиль нашёлся в сделанных на смартфоне скриншотах. Но для них забота о точности цветопередачи мало осмысленна.

Вот сделал ещё один [тестик](http://89.184.66.94/test/GammaTest.htm)
У графических файлов первой таблицы пикселы имеют одинаковые числовые значения, но сильно разные профили.
Изображения второй таблицы - чёрно-белая шахматка (серые и цветные файлы) с разными профилями.

Первый файл в каждой строке - без профиля, чтобы можно было поиграться с параметром `gfx.color_management.mode` и влиянием профиля монитора, назначенного в Панели управления Windows.
